src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+                            reportError();
+                            skipFinal = true;
+                            return;
+                        } else
+                        if (v == REPORT_COMPLETED) {
+                            child.onCompleted();
+                            skipFinal = true;
+                            return;
+                        }
+                        @SuppressWarnings("unchecked")
+                        T v2 = (T)v;
+                        child.onNext(v2);
                     }
                 }
-                if (sendOnComplete) {
-                    drainAndComplete();
+            } finally {
+                if (!skipFinal) {
+                    synchronized (this) {
+                        emitting = false;
+                    }
                 }
-            } else {
-                actual.onError(e);
             }
         }
-
-        @Override
-        public void onCompleted() {
-            boolean c = false;
+        void fastEmitError() {
             synchronized (this) {
-                completed = true;
-                if (wip == 0) {
-                    c = true;
+                if (emitting) {
+                    Object[] o = new Object[1];
+                    o[0] = REPORT_ERROR;
+                    fastArray = o;
+                    fastSize = 1;
+                    missed = true;
+                    return;
                 }
+                emitting = true;
             }
-            if (c) {
-                // complete outside of lock
-                drainAndComplete();
-            }
+            reportError();
+            // terminal event, no need to loop or stop emitting
         }
-
-        void completeInner(InnerSubscriber<T> s) {
-            boolean sendOnComplete = false;
+        void fastEmitCompleted() {
             synchronized (this) {
-                wip--;
-                if (wip == 0 && completed) {
-                    sendOnComplete = true;
+                if (emitting) {
+                    int s = fastSize;
+                    Object[] o = fastArray;
+                    if (o == null) {
+                        o = new Object[4];
+                        fastArray = o;
+                    } else
+                    if (s == o.length) {
+                        Object[] o2 = new Object[s + (s >> 2)];
+                        System.arraycopy(o, 0, o2, 0, s);
+                        fastArray = o;
+                    }
+                    o[s] = REPORT_COMPLETED;
+                    fastSize = s+ 1;
+                    missed = true;
+                    return;
                 }
+                emitting = true;
             }
-            childrenSubscribers.remove(s.sindex);
-            if (sendOnComplete) {
-                drainAndComplete();
-            }
+            // if we get here, fastArray is always null because fastEmitNext always empties the arrays
```

Comment:
It should always empty the arrays for this logic to work, but the `MAX_VALUE` for-loop does all an edge case where data could still be enqueued doesn't it?